File: mini-jdbc/src/main/java/com/mini/core/jdbc/support/Join.java
package com.mini.core.jdbc.support;

import com.mini.core.jdbc.builder.SQLBuilder;

import java.lang.annotation.*;

@Documented
@Target({ElementType.TYPE})
@Repeatable(Join.List.class)
@Retention(RetentionPolicy.RUNTIME)
public @interface Join {
    JoinType type() default JoinType.INNER;

    String value();

    enum JoinType {
        INNER {
            public void execute(SQLBuilder builder, String value) {
                builder.JOIN(value);
            }
        },
        LEFT {
            public void execute(SQLBuilder builder, String value) {
                builder.LEFT_JOIN(value);
            }
        },
        RIGHT {
            public void execute(SQLBuilder builder, String value) {
                builder.RIGHT_JOIN(value);
            }
        },
        OUTER {
            public void execute(SQLBuilder builder, String value) {
                builder.OUTER_JOIN(value);
            }
        };

        public abstract void execute(SQLBuilder builder, String value);
    }

    @Documented
    @Target({ElementType.TYPE})
    @Retention(RetentionPolicy.RUNTIME)
    @interface List {
        Join[] value() default {};
    }

}

File: mini-jdbc/src/main/kotlin/com/mini/core/jdbc/support/Join.kt
@file:Suppress("unused")

package com.mini.core.jdbc.support


import com.mini.core.jdbc.builder.SQLBuilder
import kotlin.annotation.AnnotationRetention.RUNTIME
import kotlin.annotation.AnnotationTarget.CLASS
import kotlin.annotation.AnnotationTarget.FIELD

@MustBeDocumented
@kotlin.annotation.Repeatable
@kotlin.annotation.Target(FIELD)
@kotlin.annotation.Retention(RUNTIME)
annotation class Join(val type: JoinType = JoinType.DEF, val value: String) {
    enum class JoinType {
        DEF {
            override fun execute(builder: SQLBuilder, join: String) {
                builder.JOIN(join)
            }
        },
        LEFT {
            override fun execute(builder: SQLBuilder, join: String) {
                builder.LEFT_JOIN(join)
            }
        },
        INNER {
            override fun execute(builder: SQLBuilder, join: String) {
                builder.INNER_JOIN(join)
            }
        },
        RIGHT {
            override fun execute(builder: SQLBuilder, join: String) {
                builder.RIGHT_JOIN(join)
            }
        },
        OUTER {
            override fun execute(builder: SQLBuilder, join: String) {
                builder.OUTER_JOIN(join)
            }
        };

        abstract fun execute(builder: SQLBuilder, join: String)
    }

    @MustBeDocumented
    @kotlin.annotation.Target(CLASS)
    @kotlin.annotation.Retention(RUNTIME)
    annotation class List(vararg val value: Join = [])
}